### PR TITLE
Narrow authored iframe embeds to an approved-provider allowlist

### PR DESCRIPTION
**Pull request by jeremy**
Closes the `DG4` residue: `HtmlScrubber` allowed `<iframe src>` from **any** origin in authored markdown, rendered to unauthenticated book readers.

Introduces `EmbedAllowlist` as the single source of truth for permitted embed hosts, **enforced twice**:
- **Display-time** (`HtmlScrubber`): every content render path passes through `sanitize_content`, which keeps an `<iframe>` only when its `src` host is on the allowlist and minimizes a survivor to a vetted attribute set (drop `srcdoc`, `on*`, `name`, `sandbox` overrides; filter `allow` to a safe feature set, preserving each permitted directive's authored allowlist syntax). Stored Markdown stays raw, so the list applies retroactively to existing content.
- **Browser-side** (CSP `frame-src`): consumes the same `EmbedAllowlist`, so a provider the scrubber keeps is one the browser will load — the two enforcement points **cannot drift**. Scrubber matching mirrors CSP source semantics: `*.host` covers subdomains only (never the apex), and a source narrower than a plain https origin (path, port) feeds `frame-src` verbatim while admitting no iframes.

The host set is per-install configurable via the same `CSP_EXTRA_FRAME_SRC` ENV the CSP branch tokenizes (Writebook is a self-hosted ONCE product), plus a shipped default provider list.

⚠️ **PRODUCT DECISION** (flagged in `EmbedAllowlist`): the default provider list, per-provider attribute policy, grandfathering of existing off-allowlist embeds, and any raw-iframe escape hatch are product/authoring calls an owner must confirm before this enforces.

**Stacked on** `security/xss-campaign-writebook-csp-report-only`. Tests green (embed allowlist, scrubber, CSP, pages controller). Draft spike for the XSS counteroffensive Tier C.

### CODE REVIEW

File: config/initializers/content_security_policy.rb [hunk 1]
```diff
@@ -103,6 +107,103 @@ def self.apply(policy)
...
+
+      # Extract a bare host from a CSP source expression: "https://www.youtube.com"
+      # or "https://*.vimeo.com/embed" => "www.youtube.com" / "*.vimeo.com". Returns
+      # nil for scheme-only or keyword tokens (:self, https:, 'unsafe-inline', a
+      # host:port) that carry no plain host to match an iframe src against.
+      def host_from_source(source)
+        token = source.to_s.strip.delete_prefix("https://").delete_prefix("http://")
+        token = token.split("/").first.to_s
+        return nil if token.blank? || token.include?(":") || token.start_with?("'")
+        token
```
[the remark this answers is not in the document]
Comment (jeremy) [reply config/initializers/content_security_policy.rb]: Fixed in 6ac0220 — a source expression narrower than a plain https origin (path, port) now contributes no scrubber host at all: it feeds frame-src verbatim and admits no iframes, so the scrubber can only accept what the CSP source actually allows. Wildcards now match subdomains only, never the apex. Both covered by tests.

File: config/initializers/content_security_policy.rb [hunk 2]
```diff
@@ -103,6 +107,103 @@ def self.apply(policy)
...
+        URI.parse(src.to_s.strip)
+      rescue URI::InvalidURIError
+        nil
+      end
+
+      def host_matches?(host, pattern)
+        pattern = pattern.downcase
+        if pattern.start_with?("*.")
+          apex = pattern.delete_prefix("*.")
+          host == apex || host.end_with?(".#{apex}")
```
[the remark this answers is not in the document]
Comment (jeremy) [reply config/initializers/content_security_policy.rb]: Fixed in 6ac0220 — `host_matches?` mirrors CSP host-source semantics: `*.example.test` covers subdomains only; the apex needs its own listing. The DEFAULT_PROVIDERS doc and the wildcard test flip accordingly.

File: app/models/html_scrubber.rb [hunk 3]
```diff
@@ -1,8 +1,64 @@
...
+          node[attr.name] = filtered_allow(attr.value) if name == "allow"
+        else
+          attr.remove
+        end
+      end
+    end
+
+    def filtered_allow(value)
+      value.to_s.split(";")
+        .map { |directive| directive.strip.split(/\s+/).first.to_s.downcase }
```
[the remark this answers is not in the document]
Comment (jeremy) [reply app/models/html_scrubber.rb]: Fixed in 6ac0220 — `filtered_allow` keeps each permitted directive whole rather than truncating to its feature name, so `autoplay 'none'` and `clipboard-write https://other.example` survive verbatim; disallowed features are still dropped. Test added.

File: config/initializers/content_security_policy.rb [hunk 4]
```diff
@@ -103,6 +107,103 @@ def self.apply(policy)
   end
 end
 
+# Single source of truth for which iframe embed providers Writebook permits.
+#
+# Consumed at two enforcement points that must never disagree:
+#   - author-time, by HtmlScrubber, which strips any <iframe> whose src host is
+#     not on this list before the page is stored/rendered; and
+#   - render-time, by the CSP frame-src directive below, which refuses to load a
+#     frame from an origin not on this list.
```
[the remark this answers is not in the document]
Comment (jeremy) [reply config/initializers/content_security_policy.rb]: Corrected in 6ac0220 rather than moving enforcement: display-time filtering is the intended design — every render path funnels through `sanitize_content` (pages, edits, previews, search), stored Markdown stays raw, and the allowlist therefore applies retroactively to pre-existing embeds, which storage-time enforcement can't do. The EmbedAllowlist comment and PR body now say display-time, not author/storage-time. There are no raw-content HTML consumers that bypass the helper.

**Comment by jeremy** [comment]
@codex review

File: config/initializers/content_security_policy.rb [hunk 5]
```diff
@@ -103,6 +107,113 @@ def self.apply(policy)
...
+      (default_hosts + extra_hosts).uniq
+    end
+
+    # True when +src+ is an https URL whose host is on the allowlist.
+    def allows?(src)
+      uri = parse(src)
+      return false unless uri && ALLOWED_SCHEMES.include?(uri.scheme) && uri.host.present?
+
+      host = uri.host.downcase
+      hosts.any? { |pattern| host_matches?(host, pattern) }
```
[the remark this answers is not in the document]
Comment (jeremy) [reply config/initializers/content_security_policy.rb:163]: Fixed in the follow-up commit — `allows?` now requires the default port (`uri.port == uri.default_port`), with an explicit `:443` matching like an omitted one. That closes the last dimension of the CSP host-source grammar in the scrubber matcher: scheme, host, wildcard, path, and port all now align with what frame-src will actually permit.

**Comment by jeremy** [comment]
@codex review

File: app/models/html_scrubber.rb [hunk 6]
```diff
@@ -1,8 +1,67 @@
 class HtmlScrubber < Rails::Html::PermitScrubber
+  # Attributes preserved on a surviving <iframe>. Everything else — srcdoc, on*
+  # handlers, name, sandbox overrides, arbitrary allow — is dropped so only the
+  # vetted embed shape survives. See EmbedAllowlist for the host allowlist.
+  IFRAME_ATTRIBUTES = %w[ src width height allowfullscreen loading referrerpolicy title frameborder allow ].freeze
```
[the remark this answers is not in the document]
Comment (jeremy) [reply app/models/html_scrubber.rb]: Fixed in 42c0f67 — referrerpolicy is filtered to a non-leaking allowlist (no-referrer / origin / same-origin / strict-origin / strict-origin-when-cross-origin / origin-when-cross-origin); unsafe-url and the downgrade-tolerant default are dropped. Tested both ways.

[the remark this answers is not in the document]
Comment (jeremy) [reply app/models/html_scrubber.rb]: Fixed in 42c0f67 — an authored sandbox survives minimization verbatim. You're right about the direction: the attribute can only remove privileges relative to no attribute, so stripping it was the one move a sanitizer must never make. The prior test asserting the drop flipped to assert survival.

File: config/initializers/content_security_policy.rb [hunk 7]
```diff
@@ -103,6 +107,117 @@ def self.apply(policy)
...
+      # narrower or stranger than a plain https origin — a path, a port, a
+      # scheme-only or keyword token — returns nil: stripping the qualifier
+      # would make the scrubber accept more than the CSP source actually
+      # allows, so such sources feed frame-src only and admit no iframes.
+      def host_from_source(source)
+        token = source.to_s.strip
+        return nil if token.start_with?("'")
+
+        token = token.delete_prefix("https://").delete_prefix("http://")
+        if token.blank? || token.include?("/") || token.include?(":")
```
[the remark this answers is not in the document]
Comment (jeremy) [reply config/initializers/content_security_policy.rb:216]: Fixed in 42c0f67 — an explicit :443 is normalized as the plain https origin it is (CSP matches it like an omitted port), so it feeds both frame-src and the scrubber; other ports still feed frame-src only. Tested.

File: test/controllers/pages_controller_test.rb [hunk 8]
```diff
@@ -24,10 +24,17 @@ class PagesControllerTest < ActionDispatch::IntegrationTest
...
+  test "show strips an off-allowlist iframe" do
     get leafable_path(sample_page_leaf(%(<div id="test"><iframe src="http://example.com"></iframe></div>)))
 
-    assert_select "#test", html: %(<iframe src="http://example.com"></iframe>)
+    assert_select "#test", html: %()
+    assert_select "iframe", false
+  end
+
+  test "show keeps an allowlisted-provider iframe" do
+    get leafable_path(sample_page_leaf(%(<div id="test"><iframe src="https://www.youtube.com/embed/abc"></iframe></div>)))
```
[the remark this answers is not in the document]
Comment (jeremy) [reply test/controllers/pages_controller_test.rb:35]: Not doing this — `sample_page_leaf` is this file's own pre-existing setup idiom (the neighboring tables test and others use it), and the two new tests follow it. Switching just these two to `leaves(:welcome_page)` plus a body rewrite would introduce a second setup pattern into the file while still mutating shared state per test, which is no closer to the convention's aim. Leaving unresolved for a human call.

**Comment by jeremy** [comment]
@codex review

File: config/initializers/content_security_policy.rb [hunk 9]
```diff
@@ -103,6 +107,119 @@ def self.apply(policy)
...
+      # would make the scrubber accept more than the CSP source actually
+      # allows, so such sources feed frame-src only and admit no iframes.
+      # An explicit :443 IS a plain https origin (CSP matches it like an
+      # omitted port), so it is normalized away rather than rejected.
+      def host_from_source(source)
+        token = source.to_s.strip
+        return nil if token.start_with?("'")
+
+        token = token.delete_prefix("https://").delete_prefix("http://").delete_suffix(":443")
+        if token.blank? || token.include?("/") || token.include?(":")
```
[the remark this answers is not in the document]
Comment (jeremy) [reply config/initializers/content_security_policy.rb:216]: Fixed in 9d045dd — a bare trailing slash normalizes away like the explicit :443 (a CSP root path matches every path on the host, so it IS the whole origin); deeper paths still feed frame-src only. Tested.

**Comment by jeremy** [comment]
@codex review

**Comment by jeremy** [comment]
🤖 Superseded by #480 per operator — the newer, more thorough take on the same iframe-embed allowlist (explicit host+path-shape matching, traversal/userinfo/port edge-case tests, per-provider attribute intersection). Closing in favor of #480.
